File: python/102_Binary_Tree_Level_Order_Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root:
            return []
        
        queue = deque()
        queue.append(root)
        res = []
        
        while queue:
            length = len(queue)
            count = 0
            tempList = []
            
            while count < length:
                cur = queue.popleft()
                tempList.append(cur.val)
                if cur.left:
                    queue.append(cur.left)
                if cur.right:
                    queue.append(cur.right)
                count += 1
            
            res.append(tempList)
        
        return res
        
        
        
## Two queues approach
#         if not root:
#             return []

#         res = []
        
#         queue1 = deque()
#         queue2 = deque()
        
#         queue1.append(root)
        
#         while queue1 or queue2:
#             tempList = []
            
#             while queue1:
#                 cur = queue1.popleft()
#                 if cur.left:
#                     queue2.append(cur.left)
#                 if cur.right:
#                     queue2.append(cur.right)
#                 tempList.append(cur.val)
            
#             if tempList:
#                 res.append(tempList)
            
#             tempList = []
            
#             while queue2:
#                 cur = queue2.popleft()
#                 if cur.left:
#                     queue1.append(cur.left)
#                 if cur.right:
#                     queue1.append(cur.right)
#                 tempList.append(cur.val)
            
#             if tempList:
#                 res.append(tempList)
        
#         return res
            
            
                
        
        
        

File: swift/102_binary_tree_level_order_traversal.swift
/// Definition for a binary tree node.
/// public class TreeNode {
///     public var val: Int
///     public var left: TreeNode?
///     public var right: TreeNode?
///     public init() { self.val = 0; self.left = nil; self.right = nil; }
///     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
///     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
///         self.val = val
///         self.left = left
///         self.right = right
///     }
/// }
class Solution {
  func levelOrder(_ root: TreeNode?) -> [[Int]] {
    guard let root = root else { return [] }

    var queue: [TreeNode] = [root]
    var eachLevelNodes: [[Int]] = []

    while !queue.isEmpty {
      let curLevelNodesCnt: Int = queue.count
      var curLevelNodes: [Int] = []

      for _ in 0..<curLevelNodesCnt {
        let curNode: TreeNode = queue.removeFirst()

        if curNode.left != nil {
          queue.append(curNode.left!)
        }

        if curNode.right != nil {
          queue.append(curNode.right!)
        }

        curLevelNodes.append(curNode.val)
      }

      eachLevelNodes.append(curLevelNodes)
    }

    return eachLevelNodes
  }
}
